### PR TITLE
Added reason as first-class requestable attribute of Run

### LINKED ISSUE #349: Add a first-class `reason` attribute to runs
It would be useful for each benchmark run to have a `reason` string attribute that stores a short reason why the run was kicked off (e.g. automated commit, someone asked a bot nicely on a PR, or a manual run). This would be optionally submitted by the user, along with run name and run ID. This attribute could be displayed in a column on the home page.

_Originally posted by @ElenaHenderson in https://github.com/conbench/conbench/issues/347#issuecomment-1147983741_

### PR DESCRIPTION
**Pull request by austin3dickey**
Fixes #349.

This PR adds a `reason` attribute to the Run entity. When posting benchmarks, users may now add `"run_reason": "..."` to the body, which takes the place of parsing the run name to try to figure out why a run was kicked off. This also exposes the new Reason on the app home page as a new column.

### CODE REVIEW

File: conbench/static/app.css
Comment (austin3dickey) [inline conbench/static/app.css:24]: With the additional column, depending on screen size, sometimes the `Name` column now gets cut off and wrapped to the "responsive" plus sign icon on the left side of the DataTable. I noticed that the plus sign gets uglily overlapped by the timestamp unless we get rid of this padding.

File: conbench/templates/index.html
Comment (austin3dickey) [inline conbench/templates/index.html:123]: This was hard to get "just right" for different screen sizes and I think the autoWidth is fine enough.

File: conbench/tests/api/_expected_docs.py
Comment (austin3dickey) [inline conbench/tests/api/_expected_docs.py:308]: I'm assuming it's okay that some of these are reordered; just an artifact of how swagger works

File: migrations/versions/bc0ac747cec6_add_run_reason_column.py
Comment (austin3dickey) [inline migrations/versions/bc0ac747cec6_add_run_reason_column.py]: Looking for feedback on whether this backfill code is necessary. I think it would be nice

**Comment by austin3dickey** [comment]
Also, I chose not to add `run_reason` to the compare or history apis. Let me know if you disagree

**Comment by alistaire47** [comment]
> Also, I chose not to add `run_reason` to the compare or history apis. Let me know if you disagree

I think @boshek and I might need it in the history one to separate results when generating reports, assuming this will replace the prefixing

**Comment by boshek** [comment]
> > Also, I chose not to add `run_reason` to the compare or history apis. Let me know if you disagree
> 
> I think @boshek and I might need it in the history one to separate results when generating reports, assuming this will replace the prefixing

This is a _real_ nice addition. 

I don't think we _need_ it but at least for history api, I think it would ideal to add it. We can certainly get around it by filtering for nightly elsewhere and then joining but if you are in here mucking around, it may be worth adding right now. Given that compare doesn't even return run_name, it seems safe to skip that one though.

File: migrations/versions/bc0ac747cec6_add_run_reason_column.py
Comment (jonkeane) [reply migrations/versions/bc0ac747cec6_add_run_reason_column.py]: I think it would be nice too. I'm sure @ElenaHenderson will have opinions on it! But I'm glad you included it

File: conbench/static/app.css
Comment (jonkeane) [reply conbench/static/app.css:24]: Yeah, we don't (and shouldn't) do this here, but it would be nice to make the front page slightly more responsive without resulting in a pretty unusable UI (I find the pluses, especially on mobile) not suuppper useable.

File: conbench/entities/run.py
Comment (jonkeane) [inline conbench/entities/run.py:15]: I don't think there is, but is there value in validating the reasons? Ideally there would be a relatively small number of reasons and it isn't a free-form field (that's more or less what name is for...).

Maybe describing that in the docs is good enough?

**Review by jonkeane — APPROVED** [review]
Looking good. I like what I see, though @ElenaHenderson definitely should have a look before we merge it

File: conbench/tests/api/_expected_docs.py
Comment (ElenaHenderson) [reply conbench/tests/api/_expected_docs.py:308]: Yep.

File: migrations/versions/bc0ac747cec6_add_run_reason_column.py
Comment (ElenaHenderson) [reply migrations/versions/bc0ac747cec6_add_run_reason_column.py]: Yep, we should do this.

**Review by ElenaHenderson — APPROVED** [review]
Looks great to me! Thank you!

Please note that we should update CI to send run reason in benchmark results before merging and deploying this PR.

If we deploy this PR, backfill reason column, all the new runs will have empty reason and we will need to backfill again.

**Comment by alistaire47** [comment]
> This is a _real_ nice addition.
>
> I don't think we _need_ it but at least for history api, I think it would ideal to add it. We can certainly get around it by filtering for nightly elsewhere and then joining but if you are in here mucking around, it may be worth adding right now. Given that compare doesn't even return run_name, it seems safe to skip that one though.

Actually maybe what we really want is the ability to filter by `run_reason` in the APIs, though that's beyond the scope of this story. And agreed, this PR is 😍

**Comment by jonkeane** [comment]
> what we really want is the ability to filter by run_reason in the APIs

😍 This would be fantastic, would someone mind making an issue here for that?

**Comment by alistaire47** [comment]
> 😍 This would be fantastic, would someone mind making an issue here for that?

Brief, but...sufficient? https://github.com/conbench/conbench/issues/352

File: conbench/entities/run.py
Comment (austin3dickey) [reply conbench/entities/run.py:15]: good point; I'll flesh out the docs a little more. I don't want to be prescriptive about what possible reasons could be, but I'll include a note like "runs with the same reason will probably be compared with each other so try to keep the cardinality down"

**Comment by austin3dickey** [comment]
Okay, with the advent of #352 I won't add `run_reason` to the History APIs. If we decide it's necessary in that issue we can do it then.

**Comment by austin3dickey** [comment]
When fleshing out the docs I realized I had forgotten to add `--run-reason` to the CLI. Take another look if you want!

File: migrations/versions/bc0ac747cec6_add_run_reason_column.py
Comment (austin3dickey) [reply migrations/versions/bc0ac747cec6_add_run_reason_column.py]: I decided to wait on the backfill code for now. When we merge and deploy this PR, it will add an empty Reason column. Then we can change all pipelines to start publishing reasons. And finally I'll come back and add a migration that backfills any rows with empty reasons. Thanks for the idea @ElenaHenderson .

**Pull request merged**
